contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,401 @@
+        require(validators[validator].poolModule == address(0), "ALREADY_VALIDATOR");
+        require(voteAddressToValidator[voteAddress] == address(0), "DUPLICATE_VOTE_ADDRESS");
+        require(commission.rate <= commission.maxRate, "INVALID_COMMISSION_RATE");
+        require(commission.maxChangeRate <= commission.maxRate, "INVALID_MAX_CHANGE_RATE");
+
+        // check vote address
+        _checkVoteAddress(voteAddress, blsProof);
+        voteAddressToValidator[voteAddress] = validator;
+
+        // deploy stake pool
+        address poolModule = _deployStakePool(validator);
+        totalPooledBNB += delegation;
+
+        Validator storage valInfo = validators[validator];
+        valInfo.consensusAddress = validator;
+        valInfo.poolModule = poolModule;
+        valInfo.totalPooledBNB = delegation;
+        valInfo.totalShares = delegation;
+        valInfo.voteAddress = voteAddress;
+        valInfo.description = description;
+        valInfo.commission = commission;
+
+        // update eligible validators
+        _updateEligibleValidators(validator, delegation, voteAddress);
+    }
+
+    function editDescription(Description calldata description) external onlyInitialized validatorExist(msg.sender) {
+        Validator storage valInfo = validators[msg.sender];
+        valInfo.description = description;
+    }
+
+    function editVoteAddress(bytes calldata voteAddress, bytes calldata blsProof) external onlyInitialized validatorExist(msg.sender) {
+        Validator storage valInfo = validators[msg.sender];
+        require(voteAddressToValidator[voteAddress] == address(0), "DUPLICATE_VOTE_ADDRESS");
+
+        _checkVoteAddress(voteAddress, blsProof);
+
+        voteAddressToValidator[voteAddress] = msg.sender;
+        valInfo.voteAddress = voteAddress;
+    }
+
+    function editCommissionRate(uint256 commissionRate) external onlyInitialized validatorExist(msg.sender) {
+        Validator storage valInfo = validators[msg.sender];
+        require(commissionRate <= valInfo.commission.maxRate, "INVALID_COMMISSION_RATE");
+
+        if (commissionRate > valInfo.commission.rate) {
+            require(commissionRate - valInfo.commission.rate <= valInfo.commission.maxChangeRate, "INVALID_MAX_CHANGE_RATE");
+        } else {
+            require(valInfo.commission.rate - commissionRate <= valInfo.commission.maxChangeRate, "INVALID_MAX_CHANGE_RATE");
+        }
+
+        valInfo.commission.rate = commissionRate;
+    }
+
+    function delegate(address validator) external payable onlyInitialized validatorExist(validator) {
+        uint256 _bnbAmount = msg.value;
+        require(_bnbAmount >= minDelegationChange, "INVALID_DELEGATION_AMOUNT");
+        address delegator = msg.sender;
+        Validator storage valInfo = validators[validator];
+        address pool = valInfo.poolModule;
+        if (valInfo.jailed || IStakePool(pool).getPooledBNBByShares(IStakePool(pool).balanceOf(validator)) < doubleSignSlashAmount) {
```

Comment:
I think the second check is not required. because if that happens, the validator shoule be in jailed.

Reply:
Done